File: dotnet/src/CdaToFhirAdapter.Api/Controllers/ConversionController.cs
using CdaToFhirAdapter.Core.Exceptions;
using CdaToFhirAdapter.Core.Services;
using Microsoft.AspNetCore.Mvc;

namespace CdaToFhirAdapter.Api.Controllers;

/// <summary>REST controller exposing the CDA-to-FHIR R4 conversion endpoint.</summary>
[ApiController]
[Route("api/v1")]
public sealed class ConversionController : ControllerBase
{
    private readonly ConversionService _conversionService;
    private readonly ILogger<ConversionController> _logger;

    /// <summary>Constructs the controller.</summary>
    public ConversionController(ConversionService conversionService,
        ILogger<ConversionController> logger)
    {
        _conversionService = conversionService;
        _logger = logger;
    }

    /// <summary>
    /// Converts an HL7 CDA R2 document to a FHIR R4 document Bundle.
    /// </summary>
    /// <param name="format">Output format: <c>json</c> (default) or <c>xml</c>.</param>
    /// <returns>Serialised FHIR Bundle.</returns>
    /// <response code="200">Conversion successful.</response>
    /// <response code="400">Invalid or malformed CDA document.</response>
    /// <response code="422">CDA parsed but FHIR mapping failed.</response>
    [HttpPost("convert")]
    [Consumes("application/xml", "text/xml")]
    [Produces("application/json", "application/xml")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status422UnprocessableEntity)]
    public IActionResult Convert(
        [FromBody] string cdaXml,
        [FromQuery] string format = "json")
    {
        // Sanitise user-provided format to prevent log injection
        var safeFormat = format is "json" or "xml" ? format : "json";
        _logger.LogInformation("Received CDA conversion request, format={Format}", safeFormat);

        try
        {
            var result = _conversionService.Convert(cdaXml, safeFormat);
            return Content(result.FhirBundle, result.ContentType);
        }
        catch (CdaParseException ex)
        {
            _logger.LogWarning(ex, "CDA parse error");
            return Problem(
                detail: ex.Message,
                title: "Invalid CDA document",
                statusCode: StatusCodes.Status400BadRequest);
        }
        catch (FhirMappingException ex)
        {
            _logger.LogWarning(ex, "FHIR mapping error");
            return Problem(
                detail: ex.Message,
                title: "FHIR mapping failed",
                statusCode: StatusCodes.Status422UnprocessableEntity);
        }
    }

    /// <summary>Returns a liveness probe response.</summary>
    [HttpGet("health")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    public IActionResult Health() => Ok(new { status = "healthy" });
}

File: java/src/main/java/ie/hl7/cda/fhir/api/ConversionController.java
package ie.hl7.cda.fhir.api;

import ie.hl7.cda.fhir.exception.CdaParseException;
import ie.hl7.cda.fhir.exception.FhirMappingException;
import ie.hl7.cda.fhir.model.ConversionResult;
import ie.hl7.cda.fhir.service.ConversionService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.Map;

/**
 * REST controller that exposes the CDA-to-FHIR R4 conversion endpoint.
 */
@RestController
@RequestMapping("/api/v1")
@Tag(name = "Conversion", description = "CDA to FHIR R4 conversion operations")
public class ConversionController {

    private static final Logger LOG = LoggerFactory.getLogger(ConversionController.class);

    private final ConversionService conversionService;

    /**
     * Constructs the controller.
     *
     * @param conversionService the conversion service
     */
    public ConversionController(final ConversionService conversionService) {
        this.conversionService = conversionService;
    }

    /**
     * Converts a CDA XML document to a FHIR R4 Bundle.
     *
     * @param cdaXml the CDA XML document body
     * @param format the desired output format ({@code json} or {@code xml}); defaults to {@code json}
     * @return {@code 200 OK} with the FHIR Bundle, or an error response
     */
    @PostMapping(value = "/convert",
            consumes = MediaType.APPLICATION_XML_VALUE,
            produces = {MediaType.APPLICATION_JSON_VALUE, MediaType.APPLICATION_XML_VALUE})
    @Operation(summary = "Convert CDA document to FHIR R4 Bundle",
            description = "Accepts an HL7 CDA R2 document and returns a FHIR R4 document Bundle",
            responses = {
                @ApiResponse(responseCode = "200", description = "Conversion successful",
                        content = @Content(mediaType = "application/json",
                                schema = @Schema(type = "string", format = "FHIR Bundle JSON"))),
                @ApiResponse(responseCode = "400", description = "Invalid CDA input"),
                @ApiResponse(responseCode = "422", description = "CDA parsed but FHIR mapping failed")
            })
    public ResponseEntity<String> convert(
            @RequestBody final String cdaXml,
            @Parameter(description = "Output format: json (default) or xml")
            @RequestParam(defaultValue = "json") final String format) {

        LOG.info("Received CDA conversion request, format={}", format);
        ConversionResult result = conversionService.convert(cdaXml, format);
        return ResponseEntity.ok()
                .contentType(MediaType.parseMediaType(result.getContentType()))
                .body(result.getFhirBundle());
    }

    /**
     * Global exception handler for {@link CdaParseException}.
     *
     * @param ex the exception
     * @return {@code 400 Bad Request} with the error message
     */
    @org.springframework.web.bind.annotation.ExceptionHandler(CdaParseException.class)
    public ResponseEntity<Map<String, String>> handleCdaParseException(final CdaParseException ex) {
        LOG.warn("CDA parse error: {}", ex.getMessage());
        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                .body(Map.of("error", "Invalid CDA document", "detail", ex.getMessage()));
    }

    /**
     * Global exception handler for {@link FhirMappingException}.
     *
     * @param ex the exception
     * @return {@code 422 Unprocessable Entity} with the error message
     */
    @org.springframework.web.bind.annotation.ExceptionHandler(FhirMappingException.class)
    public ResponseEntity<Map<String, String>> handleMappingException(final FhirMappingException ex) {
        LOG.warn("FHIR mapping error: {}", ex.getMessage());
        return ResponseEntity.status(HttpStatus.UNPROCESSABLE_ENTITY)
                .body(Map.of("error", "FHIR mapping failed", "detail", ex.getMessage()));
    }
}
